clap_derive/src/derives/arg_enum.rs

```diff
@@ -10,16 +10,34 @@
 
 use crate::{
     attrs::{Attrs, Name, DEFAULT_CASING, DEFAULT_ENV_CASING},
+    dummies,
     utils::Sp,
 };
 
 use proc_macro2::{Span, TokenStream};
+use proc_macro_error::abort_call_site;
 use quote::quote;
 use syn::{
-    punctuated::Punctuated, spanned::Spanned, token::Comma, Attribute, DataEnum, Ident, Variant,
+    punctuated::Punctuated, spanned::Spanned, token::Comma, Attribute, Data, DataEnum, DeriveInput,
+    Fields, Ident, Variant,
 };
 
+pub fn derive_arg_enum(input: &DeriveInput) -> TokenStream {
+    let ident = &input.ident;
+
+    dummies::arg_enum(ident);
+
+    match input.data {
+        Data::Enum(ref e) => gen_for_enum(ident, &input.attrs, e),
+        _ => abort_call_site!("`#[derive(ArgEnum)]` only supports enums"),
+    }
+}
+
 pub fn gen_for_enum(name: &Ident, attrs: &[Attribute], e: &DataEnum) -> TokenStream {
+    if !e.variants.iter().all(|v| *v.fields == Fields::Unit) {
```

Comment:
Oops, wrong deref, sorry
```suggestion
    if !e.variants.iter().all(|v| v.fields == Fields::Unit) {
```

Reply:
> error[E0369]: binary operation `==` cannot be applied to type `syn::data::Fields`
  --> clap_derive/src/derives/arg_enum.rs:37:44
   |
37 |     if !e.variants.iter().all(|v| v.fields == Fields::Unit) {
   |                                   -------- ^^ ------------ syn::data::Fields
   |                                   |
   |                                   syn::data::Fields

> error: aborting due to previous error

My original code is correct.

Reply:
It depends on `syn/extra-taits` feature. I'd vote for enabling it.

Reply:
That's for a separate discussion and another PR.

Reply:
@CreepySkeleton Do you want to enable this `extra-traits` feature and improve the derive code?